### DISCUSSION TITLE
Layers order

### DISCUSSION THREAD

**Discussion by r11d11** [devemux86/cruiser#308] [Q&A] [answered]
Another small thing.
Someone gave me the following gpx file
[Moehrendorf.zip](https://github.com/devemux86/cruiser/files/12565285/Moehrendorf.zip)

I wanted to optimize the gpx file in Cruiser. I wanted to move the shaping points.
I wondered wy shaping point 4, 6 and 7 are missing? Example see the picture,
Then I realize, the author of the GPX File set the shaping points to the exact same point then the turn instruction

The shaping points are hidden in the yellow point with the two black circle.
In my opinion it would be better if you always see the number of the shaping point. 
The yellow dots of the turning instruction ist less usefull for me. To hide the number of a shaping point is confusing when you want to move the shaping point and you only look at the map.

![ShapingPointsSmall](https://github.com/devemux86/cruiser/assets/101214119/f499c7ca-2582-473e-bbd7-08c8a1e2c87a)

**Comment by DolomitesTom** [comment]
I support this topic to put the shaping points into the front and yellow dot into the background if possible. Mentioned this also already in the past.

**Comment by devemux86** [comment]
Thanks for the suggestion.

It's an OpenGL thing, since the map is 3D, the z-order of the elements is the distance from the viewer.
For example, as we rotate the map, the nearby items appear in front.

In this example, some turn instructions have coordinates a little closer than shaping points.
(even if it is with very large double precision)

Anyway, all these are a bit technical, I will see what is possible. :slightly_smiling_face: 

---

Note that there are several workarounds:
- When tap or long press on multiple points, a dialog appears to select the layer
- You can hide the turn instructions (and other route points) in "Settings | Routes | Appearance"

**Reply by r11d11** [reply]
>You can hide the turn instructions (and other route points) in "Settings | Routes | Appearance"
I did not know this option. For now it is a good workaround.

**Reply by Ciutnik** [reply]
> You can hide the turn instructions

Specifically, place them at such a high zoom level that they appear at a much higher magnification of the map than shaping points and VIAs. This works very well 👍

**Comment by 0709wiwiwi** [comment]
hé hé I remember this one.
https://forum.kurviger.de/t/app-layer-selection-dialog/6098
It think this than has resulted in the Kurviger app offer of:
"Settings | Routes | Appearance"

**Reply by devemux86** [reply]
It is actually been mentioned before and I will check it eventually.
Just with so many features in the to-do list, it takes some time! :slightly_smiling_face:

**Comment by devemux86** [comment]
I redesigned the overlay engine in Cruiser to implement the:
- https://github.com/devemux86/cruiser/discussions/335#discussioncomment-7253756

It's optimized to use less cpu / memory / battery and have better performance.

---

I also added advanced overlay options, like setting a specific z-order of layers.
e.g. waypoints always above turn instructions and other layers

**Reply by r11d11** [reply]
I am testing Cruiser b7 and I am impressed that you added a z-order layer in such a short time. Now waypoint and shaping points are always above the turn instructions. Very good. 
But in my opinion, the waypoints and the shaping points should be always under the description.
I hope this is now easy to change. If not, it's not so important.
Here are three examples. 

Here shapingpoint 13 and waypoints are a bit lost
![Screenshot_2023-10-15-10-34-21-20_beed72cb86d752c6022398a83c94aee3 (1)](https://github.com/devemux86/cruiser/assets/101214119/475a25cf-e1f0-45a0-b6c4-509da5578c2a)

The yellow turnig instructions in the text looks irrititating
![IMG_20231015_095857](https://github.com/devemux86/cruiser/assets/101214119/b8b8993f-be92-4d95-a932-3604085146ba)

Read the name is difficult
![Screenshot_2023-10-14-20-50-54-22_beed72cb86d752c6022398a83c94aee3](https://github.com/devemux86/cruiser/assets/101214119/5968d187-3352-4741-947e-3c18bfc615c5)

**Reply by devemux86** [reply]
@r11d11 thanks for the feedback!

The z-order works with layers as a whole, symbol + label.
So in image 1 the turn instructions are set below waypoints, then their labels are below too.
(will see if this can be improved)

About the images 2 and 3, we can set the favorites in a different z-order.
Like above turn instructions? Above or below waypoints?

**Reply by devemux86** [reply]
There will be some minor improvements in the next Cruiser update.

**Reply by r11d11** [reply]
> Like above turn instructions? Above or below waypoints?

I would say above waypoints. 

I am already looking forward to the next improvement in the next update. Even it's a minor.

**Comment by devemux86** [comment]
Please check with the latest:
- https://github.com/devemux86/cruiser/discussions/329

Waypoints are still above all other markers.
We could probably change the z-order of favorites (or any group of markers).

**Reply by r11d11** [reply]
I installed Cruiser B8 and now the turn instrrctions are under the  text field. Thats looks much better 👍  Ok,, the Waypoints are still above but that is not so important. Thanks for the very quick response.

**Comment by devemux86** [comment]
I implemented an important improvement in Cruiser's redesigned overlay engine:

- Info bubbles will render on their own layer and always above all other layers

- Remember: we can now precisely define the z-order of all layers
e.g. turn instructions below, favorites in middle, waypoints above other layers

It will be available in the next Cruiser update.

<img src="https://github.com/devemux86/cruiser/assets/3484020/41ddf04c-92f4-493a-9c0e-7630f69802fb" width="50%"/>

**Reply by Ciutnik** [reply]
Since the font size of the info bubbles is no longer related to the map font size (which is good), perhaps there should be an appropriate setting? Since the last beta, these bubbles have become very tiny for me  🧐

**Reply by devemux86** [reply]
@Ciutnik can you post a screenshot?

(above image is from a 5 inches FHD device)

**Reply by Ciutnik** [reply]
It's OK on the phone, but I use an 8" armored tablet for navigation. I wouldn't like to change the system font size settings because they suit other applications.

![Screenshot_20231031_190520_Cruiser](https://github.com/devemux86/cruiser/assets/122647782/fc1dea4f-148c-4805-9da3-da5499cc7349)

**Reply by devemux86** [reply]
<img src="https://github.com/devemux86/cruiser/assets/3484020/a74628ce-9186-425f-885c-432abcf6e6c5" width="50%"/>

**Comment by devemux86** [comment]
>I implemented an important improvement in Cruiser's redesigned overlay engine:
>- Info bubbles will render on their own layer and always above all other layers
>
>- Remember: we can now precisely define the z-order of all layers
>e.g. turn instructions below, favorites in middle, waypoints above other layers

The new info bubbles are available in the latest:
- https://github.com/devemux86/cruiser/discussions/329

**Comment by devemux86** [comment, marked as answer]
New overlay engine is available in [Cruiser 3.0.34](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
